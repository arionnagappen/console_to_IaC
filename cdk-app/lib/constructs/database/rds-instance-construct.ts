import { Construct } from 'constructs';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as rds from 'aws-cdk-lib/aws-rds';
import * as kms from 'aws-cdk-lib/aws-kms';

interface DBProps {
  vpc: ec2.IVpc;
  dbSubnetGroup: rds.SubnetGroup;
  dbName: string;
  dbInstanceType: ec2.InstanceType;
  dbStorage: number;
  dbEngine: rds.IInstanceEngine;
  dbCredentials: rds.Credentials;
  dbSecurityGroup: ec2.ISecurityGroup[];
  dbMultiAZ: boolean;
}

export class MyRDSInstance extends Construct {
  public readonly rdsInstance: rds.DatabaseInstance

  constructor(scope: Construct, id: string, props: DBProps){
    super(scope, id)

    const encryptionKey = new kms.Key(this, 'RdsKmsKey', {
      enableKeyRotation: true,
      alias: 'alias/myRdsKey'
    })


    this.rdsInstance = new rds.DatabaseInstance(this, 'MyRDSInstance', {
      vpc: props.vpc,
      subnetGroup: props.dbSubnetGroup,
      databaseName: props.dbName,
      instanceType: props.dbInstanceType,
      allocatedStorage: props.dbStorage,
      engine: props.dbEngine,

      credentials: props.dbCredentials,
      securityGroups: props.dbSecurityGroup,
      publiclyAccessible: false,
      storageEncrypted: true,
      storageEncryptionKey: encryptionKey,

      deletionProtection: false,
      
      multiAz: props.dbMultiAZ
    })
  }
}