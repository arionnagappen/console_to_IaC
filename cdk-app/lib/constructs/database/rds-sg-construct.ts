import { Construct } from "constructs";
import * as ec2 from 'aws-cdk-lib/aws-ec2';

export interface DBSgProps {
  vpc: ec2.IVpc;
  appSecurityGroupId: ec2.ISecurityGroup["securityGroupId"]
}

export class DatabaseSecurityGroup extends Construct {
  public readonly dbSecurityGroup: ec2.SecurityGroup;

  constructor(scope: Construct, id: string, props: DBSgProps){
    super(scope, id)

    this.dbSecurityGroup = new ec2.SecurityGroup(this, 'DBSecurityGroup', {
      vpc: props.vpc,
      description: 'RDS Database Security Group',
      allowAllOutbound: false
    })

    // Only allow inbound traffic on port 3306
    this.dbSecurityGroup.addIngressRule(
      ec2.Peer.securityGroupId(props.appSecurityGroupId),
      ec2.Port.tcp(3306)
    )

    // Only allow outbound traffic on port 3306
    this.dbSecurityGroup.addEgressRule(
      ec2.Peer.securityGroupId(props.appSecurityGroupId),
      ec2.Port.tcp(3306)
    )
  }
}