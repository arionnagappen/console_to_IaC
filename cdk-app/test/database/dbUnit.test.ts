import * as cdk from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { DatabaseSecurityGroup } from '../../lib/constructs/database/rds-sg-construct';
import { MyRDSInstance } from '../../lib/constructs/database/rds-instance-construct';
import * as rds from 'aws-cdk-lib/aws-rds';
import * as secretsmanager from 'aws-cdk-lib/aws-secretsmanager'

// Assertion Test
test('Test DB Stack for security properties', () => {
  // Instantiates the App Construct
  const app = new cdk.App();

  // Instantiates the Stack Construct & uses the App Construct as a parameter
  const stack = new cdk.Stack(app, 'TestStack');

  // Adds a Vpc
  const vpc = new ec2.Vpc(stack, 'TestVpc', {
    maxAzs: 1,
    subnetConfiguration: [{
      name: 'test-subnet',
      subnetType: ec2.SubnetType.PRIVATE_ISOLATED
    }]
  });

  const appSG = new ec2.SecurityGroup(stack, 'TestAppSG', {vpc});

  // Security Group
  const dbSG = new DatabaseSecurityGroup(stack, 'TestDbSG', {
    vpc,
    appSecurityGroupId: appSG.securityGroupId
  });

  const mySubnetGroup = new rds.SubnetGroup(stack, 'MyDbSubnetGroup', {
    description: 'Subnet group for RDS',
    vpc: vpc,
    vpcSubnets: {
      subnetType: ec2.SubnetType.PRIVATE_ISOLATED, 
    },
    removalPolicy: cdk.RemovalPolicy.DESTROY
  });

  // Retrieve Secret
  const dbSecret = secretsmanager.Secret.fromSecretNameV2(stack, 'SecretName','rds-credentials');

  // Tell RDS use the Secret
  const myCredentials = rds.Credentials.fromSecret(dbSecret);

  new MyRDSInstance(stack, 'MyDatabase', {
    vpc: vpc,
    dbSubnetGroup: mySubnetGroup,
    dbName: 'myDatabase',
    dbInstanceType: ec2.InstanceType.of(ec2.InstanceClass.T3, ec2.InstanceSize.MICRO),
    dbStorage: 10,
    dbEngine: rds.DatabaseInstanceEngine.mysql({version: rds.MysqlEngineVersion.VER_8_4_5}),
    dbCredentials: myCredentials,
    dbSecurityGroup: [dbSG.dbSecurityGroup],
    dbMultiAZ: true,
  });

  const template = Template.fromStack(stack);

  template.hasResourceProperties('AWS::RDS::DBInstance', {
    MasterUsername: Match.anyValue(),
    MasterUserPassword: Match.anyValue(),
    VPCSecurityGroups: Match.anyValue(),
    StorageEncrypted: true
  })

})