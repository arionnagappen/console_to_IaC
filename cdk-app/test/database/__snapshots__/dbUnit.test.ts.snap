// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DB Unit Test Matches SnapShot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "MyDatabaseMyRDSInstanceAA7465FD": {
      "DeletionPolicy": "Snapshot",
      "Properties": {
        "AllocatedStorage": "10",
        "CopyTagsToSnapshot": true,
        "DBInstanceClass": "db.t3.micro",
        "DBName": "myDatabase",
        "DBSubnetGroupName": {
          "Ref": "MyDbSubnetGroup",
        },
        "DeletionProtection": false,
        "Engine": "mysql",
        "EngineVersion": "8.4.5",
        "KmsKeyId": {
          "Fn::GetAtt": [
            "MyDatabaseRdsKmsKeyDE61EEDB",
            "Arn",
          ],
        },
        "MasterUserPassword": {
          "Fn::Join": [
            "",
            [
              "{{resolve:secretsmanager:arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":secretsmanager:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":secret:rds-credentials:SecretString:password::}}",
            ],
          ],
        },
        "MasterUsername": {
          "Fn::Join": [
            "",
            [
              "{{resolve:secretsmanager:arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":secretsmanager:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":secret:rds-credentials:SecretString:username::}}",
            ],
          ],
        },
        "MultiAZ": true,
        "PubliclyAccessible": false,
        "StorageEncrypted": true,
        "StorageType": "gp2",
        "VPCSecurityGroups": [
          {
            "Fn::GetAtt": [
              "TestDbSGDBSecurityGroupE28CDCBD",
              "GroupId",
            ],
          },
        ],
      },
      "Type": "AWS::RDS::DBInstance",
      "UpdateReplacePolicy": "Snapshot",
    },
    "MyDatabaseRdsKmsKeyAlias61DB8650": {
      "Properties": {
        "AliasName": "alias/myRdsKey",
        "TargetKeyId": {
          "Fn::GetAtt": [
            "MyDatabaseRdsKmsKeyDE61EEDB",
            "Arn",
          ],
        },
      },
      "Type": "AWS::KMS::Alias",
    },
    "MyDatabaseRdsKmsKeyDE61EEDB": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "EnableKeyRotation": true,
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "MyDbSubnetGroup": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "DBSubnetGroupDescription": "Subnet group for RDS",
        "SubnetIds": [
          {
            "Ref": "TestVpctestsubnetSubnet1Subnet212A675B",
          },
        ],
      },
      "Type": "AWS::RDS::DBSubnetGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "SecretNameAttachment5029B42C": {
      "Properties": {
        "SecretId": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":secretsmanager:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":secret:rds-credentials",
            ],
          ],
        },
        "TargetId": {
          "Ref": "MyDatabaseMyRDSInstanceAA7465FD",
        },
        "TargetType": "AWS::RDS::DBInstance",
      },
      "Type": "AWS::SecretsManager::SecretTargetAttachment",
    },
    "TestAppSG78644F93": {
      "Properties": {
        "GroupDescription": "TestStack/TestAppSG",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "TestDbSGDBSecurityGroupE28CDCBD": {
      "Properties": {
        "GroupDescription": "RDS Database Security Group",
        "SecurityGroupEgress": [
          {
            "Description": {
              "Fn::Join": [
                "",
                [
                  "from ",
                  {
                    "Fn::GetAtt": [
                      "TestAppSG78644F93",
                      "GroupId",
                    ],
                  },
                  ":3306",
                ],
              ],
            },
            "DestinationSecurityGroupId": {
              "Fn::GetAtt": [
                "TestAppSG78644F93",
                "GroupId",
              ],
            },
            "FromPort": 3306,
            "IpProtocol": "tcp",
            "ToPort": 3306,
          },
        ],
        "SecurityGroupIngress": [
          {
            "Description": {
              "Fn::Join": [
                "",
                [
                  "from ",
                  {
                    "Fn::GetAtt": [
                      "TestAppSG78644F93",
                      "GroupId",
                    ],
                  },
                  ":3306",
                ],
              ],
            },
            "FromPort": 3306,
            "IpProtocol": "tcp",
            "SourceSecurityGroupId": {
              "Fn::GetAtt": [
                "TestAppSG78644F93",
                "GroupId",
              ],
            },
            "ToPort": 3306,
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "TestVpcE77CE678": {
      "Properties": {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/TestVpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "TestVpctestsubnetSubnet1RouteTableA3F3FD85": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/TestVpc/test-subnetSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "TestVpctestsubnetSubnet1RouteTableAssociation9E8E56C2": {
      "Properties": {
        "RouteTableId": {
          "Ref": "TestVpctestsubnetSubnet1RouteTableA3F3FD85",
        },
        "SubnetId": {
          "Ref": "TestVpctestsubnetSubnet1Subnet212A675B",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "TestVpctestsubnetSubnet1Subnet212A675B": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/16",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "test-subnet",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          {
            "Key": "Name",
            "Value": "TestStack/TestVpc/test-subnetSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
